executor/cte_storage.go

```diff
@@ -0,0 +1,128 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"errors"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+var (
+	_ CTEStorage = &CTEStorageRC{}
+)
+
+type CTEStorage interface {
+	// If is first called, will open underlying storage. Otherwise will add ref count by one
+	// bytesLimit < 0 means no memory limit
+	OpenAndRef(fieldType []*types.FieldType, chkSize int, bytesLimit int64) error
+
+	// Minus ref count by one, if ref count is zero, close underlying storage
+	DerefAndClose() (isClosed bool, err error)
+
+	// Swap data of two storage
+	Swap(other CTEStorage) error
+
+	// Check whether storage is valid or not
+	Valid() bool
+
+	// Data will first resides in RAM, when exceeds bytesLimit, will spill all data to disk automatically
+	Add(chk *chunk.Chunk) error
+
+	GetChunk(chkIdx int) (*chunk.Chunk, error)
+}
+
+// CTEStorage implementation using RowContainer
+type CTEStorageRC struct {
+	refCnt        int
+	bytesLimit    int64
+	bytesConsumed int64
+	rc            *chunk.RowContainer
```

Comment:
We also add lock for CTEStorage, because UnionExec will concrrently call children's Next()